satellite/metainfo/loop.go

```diff
@@ -0,0 +1,214 @@
+		}
+	}
+}
+
+// Join will join the looper for one full cycle until completion and then returns.
+// On ctx cancel the observer will return without completely finishing.
+// Only on full complete iteration it will return nil.
+func (service *Loop) Join(ctx context.Context, observer Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	obsContext := &observerContext{
+		Observer: observer,
+		ctx:      ctx,
+		done:     make(chan error),
+	}
+
+	select {
+	case <-service.done:
+		return context.Canceled
+	case service.join <- obsContext:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return obsContext.Wait()
+}
+
+func (service *Loop) runOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var observers []*observerContext
+
+	defer func() {
+		for _, observer := range observers {
+			observer.Finish()
+		}
+	}()
+
+	select {
+	case observer := <-service.join:
+		observers = append(observers, observer)
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	timer := time.NewTimer(service.config.CoalesceDuration)
+waitformore:
+	for {
+		select {
+		case observer := <-service.join:
+			observers = append(observers, observer)
+		case <-timer.C:
+			break waitformore
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	return service.metainfo.Iterate(ctx, "", "", true, false,
+		func(ctx context.Context, it storage.Iterator) error {
+			var item storage.ListItem
+
+			// iterate over every segment in metainfo
+			for it.Next(ctx, &item) {
+				pointer := &pb.Pointer{}
+
+				err = proto.Unmarshal(item.Value, pointer)
+				if err != nil {
+					// TODO: figure out what to do
+					// return LoopError.New("error unmarshalling pointer %s", err)
+					continue
+				}
+
+				path := item.Key.String()
+				pathElements := storj.SplitPath(path)
+				isLastSeg := len(pathElements) >= 2 && pathElements[1] == "l"
+
+				nextObservers := observers[:0]
+
+				// send segment info to every observer
+				for _, observer := range observers {
+					remote := pointer.GetRemote()
+					if remote != nil {
+						if observer.HandleError(observer.RemoteSegment(ctx, path, pointer)) {
+							continue
+						}
+
+						if isLastSeg {
+							if observer.HandleError(observer.RemoteObject(ctx, path, pointer)) {
+								continue
+							}
+						}
+					} else if observer.HandleError(observer.InlineSegment(ctx, path, pointer)) {
+						continue
+					}
+
+					select {
+					case <-observer.ctx.Done():
+						observer.HandleError(observer.ctx.Err())
+						continue
+					default:
+					}
+
+					nextObservers = append(nextObservers, observer)
+				}
+
+				observers = nextObservers
+				if len(observers) == 0 {
+					return nil
+				}
+
+				select {
+				case <-ctx.Done():
+					for _, observer := range observers {
+						observer.HandleError(ctx.Err())
+					}
+					observers = nil
+					return ctx.Err()
+				default:
+				}
+			}
+			return nil
+		})
+}
+
+// Close halts the metainfo loop
+func (service *Loop) Close() error {
+	return nil
```

Comment:
I tried to implement it but it didn't end up being so simple. I am not sure how important implementing this would be, since I don't think anyone would use it

Reply:
implemented now